logic/source_map.go

```diff
@@ -0,0 +1,132 @@
+package logic
+
+import (
+	"fmt"
+	"strings"
+)
+
+// SourceMap provides a mapping of the source to assembled program
+type SourceMap struct {
+	Version    int      `json:"version"`
+	File       string   `json:"file,omitempty"`
+	SourceRoot string   `json:"sourceRoot,omitempty"`
+	Sources    []string `json:"sources"`
+	Names      []string `json:"names"`
+	Mappings   string   `json:"mappings"`
+	// Decoded mapping results
+	LineToPc map[int][]int
+	PcToLine map[int]int
+}
+
+func DecodeSourceMap(ism map[string]interface{}) (SourceMap, error) {
+	sm := SourceMap{}
+
+	if v, ok := ism["version"]; ok {
+		sm.Version = int(v.(float64))
+	}
+
+	if sm.Version != 3 {
+		return sm, fmt.Errorf("only version 3 is supported")
+	}
+
+	if f, ok := ism["file"]; ok {
+		sm.File = f.(string)
+	}
+
+	if sr, ok := ism["sourceRoot"]; ok {
+		sm.SourceRoot = sr.(string)
+	}
+
+	if srcs, ok := ism["sources"]; ok {
+		srcSlice := srcs.([]interface{})
+		sm.Sources = make([]string, len(srcSlice))
+		for idx, s := range srcSlice {
+			sm.Sources[idx] = s.(string)
+		}
+	}
+
+	if names, ok := ism["names"]; ok {
+		nameSlice := names.([]interface{})
+		sm.Names = make([]string, len(nameSlice))
+		for idx, n := range nameSlice {
+			sm.Names[idx] = n.(string)
+		}
+	}
+
+	if m, ok := ism["mappings"]; ok {
+		sm.Mappings = m.(string)
+	}
+
+	if sm.Mappings == "" {
+		return sm, fmt.Errorf("no mappings defined")
+	}
+
+	sm.PcToLine = map[int]int{}
+	sm.LineToPc = map[int][]int{}
+
+	lastLine := 0
+	for idx, chunk := range strings.Split(sm.Mappings, ";") {
+		vals := decodeSourceMapLine(chunk)
+		// If the vals length >= 3 the lineDelta
+		if len(vals) >= 3 {
+			lastLine = lastLine + vals[2] // Add the line delta
+		}
+
+		if _, ok := sm.LineToPc[lastLine]; !ok {
+			sm.LineToPc[lastLine] = []int{}
+		}
+
+		sm.LineToPc[lastLine] = append(sm.LineToPc[lastLine], idx)
+		sm.PcToLine[idx] = lastLine
+	}
+
+	return sm, nil
+}
+
+func (s *SourceMap) GetLineForPc(pc int) int {
```

Comment:
@barnjamin Friendly visual reminder to implement an analog of https://github.com/algorand/py-algorand-sdk/pull/353#discussion_r910177370 here.

Reply:
@barnjamin You're correct that the logic won't panic.  When I wrote my earlier note, I forgot Go's behavior is to return the zero value for a non-existent key (https://go.dev/tour/moretypes/22).

My updated thinking is:
* `GetPcsForLine` requires no change because the zero value is empty array.
* `GetLineForPc` _should_ be changed to disambiguate the zero value.  0 can mean a valid mapping or the value returning when key does not exist.
  * I'm _assuming_ the preferred Go convention here is to return `int, bool` so that the caller can validate key presence.

Reply:
@barnjamin Thanks for the change - resolving.